primitives/src/ticker.rs

```diff
@@ -0,0 +1,48 @@
+//! Ticker symbol
+use codec::{Decode, Encode};
+
+const TICKER_LEN: usize = 12;
+
+/// Ticker symbol.
+#[derive(Decode, Encode, Clone, Copy, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Ticker(pub [u8; TICKER_LEN]);
+
+impl Default for Ticker {
+    fn default() -> Self {
```

Comment:
perhaps also define a `new([0u8; TICKER_LEN])` that converts lower case to uppercase. Basically, we want it to always be upper case. Is there an idiomatic way to do it?

Alternatively, I think we can implement Eq and PartialEq to do case insensitive equality and return "aAa" === "AAa".

Reply:
If we could make `Ticker::new` available to JS and hide the array field that would be the best but that wouldn't work with the Polkadot UI, I reckon.

I think the best option practically is simply to only input uppercase strings when creating ticker symbols.

There is a way to go from arbitrary case to upper case in the constructor. But that would mean that JS calls would use byte strings.

Redefining the instances would mean redefining ordering and hashing too.

Reply:
Either always keeping tickers in upper case or always doing case insensitive equality is a hard requirement for us.

This is because functions on the blockchains can be directly called without the UI (or custom UI) and the functions/system assumes "aAa" === "AAa".

Our current solution is to basically call bytes_to_upper in every function. The main aim of this task is to avoid needing to do that. I am not sure if it's practically feasible or not though. Perhaps we can have a "transform" member function in ticker and always call that in all functions. It's not much better than using bytes_to_upper though.

Reply:
The best way to ensure uppercase representation would be to convert the contained byte string in the interface. Between the schema type and the Rust type, if that is possible. `bytes_to_upper` was probably a way to approximate this.

Reply:
There is a "transform" function - `canonize`.

Reply:
Yeah, but we aren't using it right now. If we don't find a better solution, we need to use it everywhere.

Reply:
We are using it in the places where we had `bytes_to_upper` and where the ticker was passed by value.

There were uses of `bytes_to_upper` in methods where `ticker` would be passed by reference but they all turned out to be redundant since those methods are called after a previous call to `bytes_to_upper` in a method to which `ticker` is passed by value.

Reply:
There are places where it's missing (maybe we were missing bytes_to_upper as well at those places). For example, in new function in dividend.rs, `payout_ticker` is not being canonized. In the same file, in the `cancel` function, `ticker` is not being canonized. Please make sure all values of tickers are being canonized.